### PR TITLE
playing with tut [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by mossprescott**
This is more an experiment than anything else, but I do think the results would be helpful as a reference. Also, trying to make this look nice really pointed out some of the rough edges of the API :disappointed: 

The question is where to put the output where anyone will see it. 

Here's the output for your amusement:
# Pathy Examples
## Constructing paths

Typically, you want to import both `Path` and all of the helpers in its companion:

``` scala
import pathy.Path, Path._
```

To create paths, use either `rootDir or`currentDir`, and then add segments with`</>`,`dir`, and`file`.

``` scala

rootDir </> file("afile.txt")

currentDir </> dir("somedir")
```

Note that the resulting type tracks the kind of path:

``` scala
val p1: Path[Abs, File, Sandboxed] = rootDir </> file("afile.txt")
val p2: Path[Rel, Dir, Sandboxed] = currentDir </> dir("somedir")
```

The special operator `<::>` builds paths that refer to parent directories. Note that the resulting path is always `Unsandboxed`, meaning you can't be sure that it refers to a file inside the current directory:

``` scala
val p3: Path[Rel, Dir, Unsandboxed] = currentDir <::> dir("somedir")
```

If the types don't match, you get a _compile time_ error:

``` scala
scala> val impossible: Path[Rel, Dir, Sandboxed] = rootDir </> file("afile.txt")
<console>:15: error: type mismatch;
 found   : pathy.Path[pathy.Path.Abs,pathy.Path.File,Nothing]
 required: pathy.Path[pathy.Path.Rel,pathy.Path.Dir,pathy.Path.Sandboxed]
val impossible: Path[Rel, Dir, Sandboxed] = rootDir </> file("afile.txt")
                                                    ^
```

A `Show` instance is provided, which renders paths as they would be entered in source code:

``` scala
import scalaz._, Scalaz._
```

``` scala
scala> p1.shows
res4: String = rootDir </> file("afile.txt")
```

Finally, type aliases are defined for the four interesting combinations, and you can often ignore sandboxing:

``` scala
scala> val p1a: AbsFile[_] = p1
p1a: pathy.Path.AbsFile[_] = FileIn(Root,FileName(afile.txt))
```
## Parsing paths

A _codec_ provides simple parsing and printing of paths. A couple of simple codecs are provided:

``` scala
val p4: Option[Path[Abs, Dir, Unsandboxed]] = posixCodec.parseAbsDir("/")
val p5: Option[Path[Rel, File, Unsandboxed]] = windowsCodec.parseRelFile(".\\My Documents\\bob.bmp")
```

If you try to parse a path of the wrong type, pathy lets you know:

``` scala
scala> posixCodec.parseRelFile("/actuallyADir/")
res5: Option[pathy.Path.RelFile[pathy.Path.Unsandboxed]] = None
```

To print a path, it first needs to be "sandboxed", that is, made relative to some reference path:

``` scala
scala> for {
     |   p <- posixCodec.parseRelFile("./foo/bar/baz.txt")
     |   p <- sandbox(currentDir </> dir("foo"), p)
     | } yield posixCodec.printPath(p)
res6: Option[String] = Some(./bar/baz.txt)
```
## Working with paths

_TODO: examples for the mundane stuff_

If you know you're dealing with a File path, some additional functions are available:

``` scala
val p6 = rootDir </> file("file.txt")
```

``` scala
scala> renameFile(p6, name => FileName(name.value.toUpperCase)).shows
res7: String = rootDir </> file("FILE.TXT")

scala> (p1 <:> "jpg").shows
res8: String = rootDir </> file("afile.jpg")
```

A file path _always_ has both a parent direcory and a file name at the end:

``` scala
scala> fileParent(p6)
res9: pathy.Path[pathy.Path.Abs,pathy.Path.Dir,Nothing] = Root

scala> fileName(p6)
res10: pathy.Path.FileName = FileName(file.txt)
```

On the other hand, a directory path may or may not actually contain any directory names:

``` scala
scala> dirName(currentDir)
res11: Option[pathy.Path.DirName] = None
```

### CODE REVIEW

**Comment by sellout** [comment]
I think this is cool. I’ve been meaning to play with tut, but haven’t yet. I’ll look at it a bit more tomorrow.

File: core/build.sbt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+tutSettings
+
+tutScalacOptions := Seq(
+  "-deprecation",
+  "-encoding", "UTF-8",
+  "-feature",
+  "-language:existentials",
+  "-language:higherKinds",
+  "-unchecked",
+  //"-Xfatal-warnings",  // Note: making unused imports an error breaks tut
```
Comment (sellout) [inline core/build.sbt]: You should open an issue against tut. I was about to, but then I wasn’t sure where these settings take effect, so I couldn’t describe it appropriately.

Also, could we use `tutScalacOptions = compilerOptions -- Seq("-Xfatal-warnings", …)`? (and add as many missing options to `compilerOptions` as possible)

We should really write a SlamData sbt plugin that has all of our restrictive stuff (compiler options, warts, etc.) so we don’t have to duplicate that in each project.

**Comment by mossprescott** [comment]
Is there value in leaving this open? Or some path to getting it merged? I'm OK either way, or I could have a go at a similar PR for matryoshka, if that's a better test case.

**Comment by sellout** [comment]
So, I think this seems useful – the question is where/how to publish? https://pages.github.com seems like a good option, but it requires this stuff to be on a separate branch, so I don’t know how that’d work.

What if you use tut + sbt-site, which outputs to target/site, then we can commit that to the gh-pages branch?

**Comment by mossprescott** [comment]
https://github.com/sbt/sbt-ghpages also seems helpful. I'll give this a try soon.

**Comment by mossprescott** [comment]
OK, this is now building a somewhat legit github-pages site, following stew/indoctrinate. The design is weak and lots of things could be improved, but it seems basically functional.

See https://mossprescott.github.io/scala-pathy/.

The ghpages-update plugin is in place but does not work if I try to deploy to my fork. That might start working when this gets to master, I'm not sure.

Also I'm not sure how this will play with the main https://slamdata.github.io/ page, which currently forwards to slamdata.com.

**Comment by wemrysi** [comment]
@mossprescott Looks good!

**Pull request closed without merging**